pkg/sql/config.go

```diff
@@ -26,39 +30,69 @@ func init() {
 	config.ZoneConfigHook = GetZoneConfig
 }
 
-// GetZoneConfig returns the zone config for the object with 'id'.
-func GetZoneConfig(cfg config.SystemConfig, id uint32) (config.ZoneConfig, bool, error) {
+func getZoneConfig(
+	id uint32,
+	get func(roachpb.Key) (*roachpb.Value, error),
+) (config.ZoneConfig, bool, error) {
 	// Look in the zones table.
-	if zoneVal := cfg.GetValue(sqlbase.MakeZoneKey(sqlbase.ID(id))); zoneVal != nil {
+	if zoneVal, err := get(sqlbase.MakeZoneKey(sqlbase.ID(id))); err != nil {
+		return config.ZoneConfig{}, false, err
+	} else if zoneVal != nil {
 		// We're done.
 		zone, err := config.MigrateZoneConfig(zoneVal)
 		return zone, true, err
 	}
 
 	// No zone config for this ID. We need to figure out if it's a database
 	// or table. Lookup its descriptor.
-	if descVal := cfg.GetValue(sqlbase.MakeDescMetadataKey(sqlbase.ID(id))); descVal != nil {
+	if descVal, err := get(sqlbase.MakeDescMetadataKey(sqlbase.ID(id))); err != nil {
+		return config.ZoneConfig{}, false, err
+	} else if descVal != nil {
 		// Determine whether this is a database or table.
 		var desc sqlbase.Descriptor
 		if err := descVal.GetProto(&desc); err != nil {
 			return config.ZoneConfig{}, false, err
 		}
 		if tableDesc := desc.GetTable(); tableDesc != nil {
 			// This is a table descriptor. Lookup its parent database zone config.
-			return GetZoneConfig(cfg, uint32(tableDesc.ParentID))
+			return getZoneConfig(uint32(tableDesc.ParentID), get)
 		}
 	}
 
 	// Retrieve the default zone config, but only as long as that wasn't the ID
 	// we were trying to retrieve (avoid infinite recursion).
 	if id != keys.RootNamespaceID {
-		return GetZoneConfig(cfg, keys.RootNamespaceID)
+		return getZoneConfig(keys.RootNamespaceID, get)
 	}
 
 	// No descriptor or not a table.
 	return config.ZoneConfig{}, false, nil
 }
 
+// GetZoneConfig returns the zone config for the object with 'id' using the
+// cached system config.
+func GetZoneConfig(cfg config.SystemConfig, id uint32) (config.ZoneConfig, bool, error) {
+	return getZoneConfig(id, func(key roachpb.Key) (*roachpb.Value, error) {
+		return cfg.GetValue(key), nil
+	})
+}
+
+// GetZoneConfigInTxn is like GetZoneConfig, but uses the provided transaction
+// to perform lookups instead of the cached system config.
+func GetZoneConfigInTxn(
```

Comment:
When would this be used? We call `GetZoneConfig` **a lot** within pretty tight loops (think the various queues).
Is the plan to stop gossiping the zone configs, or parts of it?

Reply:
Certainly not planning to stop gossiping the zone config! Rather I'm moving parts of the `./cockroach zone` CLI into SQL, and I want to be able to read/update a zone config within a transaction. Basically, I'm implementing

```
ALTER TABLE db.tbl EXPERIMENTAL SET ZONE 'some yaml'
```

The CLI currently accomplishes this by running:

```
BEGIN;
SELECT config FROM system.zones WHERE id = 51;
--If missing...
SELECT config FROM system.zones WHERE id = 50;
--If still missing...
SELECT config FROM system.zones WHERE id = 0;
UPDATE system.zones SET config = 'some derived config' WHERE id = 51;
COMMIT;
```

(That PR should be out for review later today.)

Reply:
Ah, so it's purely for settings purposes. Consumption of the zone configs would still be through the gossiped (and cached) version? If so, this sees reasonable.

Reply:
Yep, exactly!